### PR TITLE
Test that push doesn't work for local non-bare repos

### PR DESCRIPTION
**Pull request by kshyatt**

### CODE REVIEW

**Review by ararslan — APPROVED** [review]

**Comment by kshyatt** [comment]
`threads` segfault seems unrelated?

**Comment by ararslan** [comment]
Yes

**Comment by tkelman** [comment]
Possibly not, I see
```
git_config_free at /tmp/julia/bin/../lib/julia/libgit2.so (unknown line)
close at ./libgit2/types.jl:547
```
in part of the backtrace

**Comment by kshyatt** [comment]
@tkelman fixed?

**Comment by tkelman** [comment]
or intermittent, but probably good enough

File: test/libgit2.jl [hunk 1]
```diff
@@ -790,6 +790,32 @@ mktempdir() do dir
...
+        LibGit2.set!(up_cfg, "user.name", "AAAA")
+        LibGit2.set!(up_cfg, "user.email", "BBBB@BBBB.COM")
+        try
+            open(joinpath(LibGit2.path(our_repo),"file1"),"w") do f
+                write(f, "111\n")
+            end
+            LibGit2.add!(our_repo, "file1")
+            LibGit2.commit(our_repo, "add file1")
+            # we cannot yet locally push to non-bare repos
+            @test_throws LibGit2.GitError LibGit2.push(our_repo, remoteurl=up_path)
```
Comment (tkelman) [inline test/libgit2.jl]: is this intended roughly as a `@test_broken`, should work but doesn't, fixme?

Comment (kshyatt) [inline test/libgit2.jl]: No, I think this is a `libgit2` issue. It's not a problem on our end.

Comment (tkelman) [inline test/libgit2.jl]: Worth reporting then. Do you know which ccall is causing the problem?

**Comment by tkelman** [comment]
There's definitely a segfault that I can get to happen locally (on linux at least) from running the libgit2 test repeatedly. Not sure whether it was introduced by this PR or one of the other related ones, but will try to narrow it down.

**Comment by nalimilan** [comment]
Unfortunately, this test triggers an assertion failure inside libgit2 when using the system version of the library, which does not include the patch from https://github.com/JuliaLang/julia/issues/21639. This is annoying for packagers and goes against the claim that we support libgit2 ≥ 0.23 (or even 0.25.1).

I'm not sure how to handle this. We could run this test only when libgit2 is more recent than 0.25.1, but that would mean the test would always be disabled for now. Is there a way to detect that the patched version is used? Or maybe we could use a custom build number which would be considered as higher than 0.25.1?

**Comment by tkelman** [comment]
if feasible, patching the build number to something custom when we know our build has the bug fix included sounds okay to me

**Comment by tkelman** [comment]
I'll be opening a branch for rc2 backports shortly, any ideas for what would be doable here?

**Comment by nalimilan** [comment]
Unfortunately, looking at the code it appears that `LibGit2.version` cannot return the build number, as the underlying libgit2 function only returns `major.minor.patch`. So I'd say the best solution for now is to disable the test with libgit2 <= 0.25.1. See https://github.com/JuliaLang/julia/pull/21872.

**Pull request merged**